whois-api/src/main/java/net/ripe/db/whois/api/nrtm4/NrtmClientService.java

```diff
@@ -0,0 +1,136 @@
+                             final SnapshotSourceAwareFileRepository snapshotSourceAwareFileRepository,
+                             final DeltaSourceAwareFileRepository deltaSourceAwareFileRepository) {
+        this.snapshotSourceAwareFileRepository = snapshotSourceAwareFileRepository;
+        this.deltaSourceAwareFileRepository = deltaSourceAwareFileRepository;
+        this.notificationFileSourceAwareDao = notificationFileSourceAwareDao;
+        this.sourceRepository = sourceRepository;
+        this.nrtmUrl = nrtmUrl;
+    }
+
+    @GET
+    @Produces(MediaType.TEXT_HTML)
+    public String sourcesLinkAsHtml() {
+
+        final StringBuilder sourceLink = new StringBuilder();
+
+        sourceRepository.getAllSources().forEach(sourceModel ->
+                sourceLink.append(
+                        String.format("<a href='%s'>%s</a><br/>",
+                                        String.join("/", nrtmUrl, sourceModel.getName().toString(), "update-notification-file.json"),
+                                        sourceModel.getName().toString()
+                                    )
+                )
+        );
+
+        return String.format(SOURCE_LINK_PAGE, sourceLink);
+    }
+
+    @GET
+    @Path("{source}/{filename}")
+    @Produces({MediaType.APPLICATION_OCTET_STREAM, MediaType.APPLICATION_JSON})
+    public Response nrtmFiles(
+            @Context final HttpServletRequest httpServletRequest,
+            @PathParam("source") final String source,
+            @PathParam("filename") final String fileName) {
+
+        if(fileName.startsWith(NrtmDocumentType.NOTIFICATION.getFileNamePrefix())) {
+            return notificationFileSourceAwareDao.findLastNotification(getSource(source))
+                    .map( payload -> getResponse(payload,  NrtmFileUtil.NOTIFICATION_FILENAME))
+                    .orElseThrow(() -> new NotFoundException("update-notification-file.json does not exists for source " + source));
+        }
+
+        validateSource(source, fileName);
+        if(fileName.startsWith(NrtmDocumentType.SNAPSHOT.getFileNamePrefix())) {
+            return snapshotSourceAwareFileRepository.getByFileName(fileName)
+                    .map( snapshot -> getResponse(snapshot.getPayload(), snapshot.getSnapshotFile().hash(), fileName))
+                    .orElseThrow(() -> new NotFoundException("Requested Snapshot file does not exists"));
+        }
+
+        final String filenameExt  = filenameWithExtension(fileName);
+        if(fileName.startsWith(NrtmDocumentType.DELTA.getFileNamePrefix())) {
+            return deltaSourceAwareFileRepository.getByFileName(filenameExt)
+                    .map( delta -> getResponse(delta.payload(), delta.hash(), filenameExt))
+                    .orElseThrow(() -> new NotFoundException("Requested Delta file does not exists"));
+        }
+
+        throw new BadRequestException("Invalid Nrtm filename");
```

Comment:
Return 404 Not Found instead?

Reply:
Not really it is invalid file now , as it does not match to any of snapshot/delta or notification file syntax